Source/SantaGUI/SNTNotificationManager.m

```diff
@@ -143,4 +165,74 @@ - (void)postRuleSyncNotificationWithCustomMessage:(NSString *)message {
   [[NSUserNotificationCenter defaultUserNotificationCenter] deliverNotification:un];
 }
 
+#pragma mark SNTBundleNotifierXPC protocol methods
+
+- (void)updateTotalFileCountForEvent:(SNTStoredEvent *)event
+                         binaryCount:(uint64_t)binaryCount
+                           fileCount:(uint64_t)fileCount {
+  if ([self.currentWindowController.event.idx isEqual:event.idx]) {
+    dispatch_async(dispatch_get_main_queue(), ^{
+      self.currentWindowController.foundFileCountLabel.stringValue =
+          [NSString stringWithFormat:@"%llu binaries / %llu files", binaryCount, fileCount];
+    });
+  }
+}
+
+- (void)setBundleServiceListener:(NSXPCListenerEndpoint *)listener {
+  SNTXPCConnection *c = [[SNTXPCConnection alloc] initClientWithListener:listener];
+  c.remoteInterface = [SNTXPCBundleServiceInterface bundleServiceInterface];
+  [c resume];
+  self.bundleServiceConnection = c;
+  dispatch_semaphore_signal(self.bundleServiceSema);
+}
+
+#pragma mark SNTBundleNotifierXPC helper methods
+
+- (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event {
+  // Wait until the bundle service is connected
+  dispatch_semaphore_wait(self.bundleServiceSema, DISPATCH_TIME_FOREVER);
+
+  // Let all future requests flow, until the connection is terminated and we go back to waiting.
+  dispatch_semaphore_signal(self.bundleServiceSema);
+
+  // NSProgress becomes current for this thread. XPC messages vend a child node to the receiver.
+  [self.currentWindowController.progress becomeCurrentWithPendingUnitCount:1];
+
+  // Start hashing. Progress is reported to the root NSProgress (currentWindowController.progress).
+  [[self.bundleServiceConnection remoteObjectProxy]
+      hashBundleBinariesForEvent:event
+                           reply:^(NSString *bh, NSArray<SNTStoredEvent *> *events, NSNumber *ms) {
+     if (!bh) return;
+     event.fileBundleHash = bh;
+     event.fileBundleBinaryCount = @(events.count);
+     event.fileBundleHashMilliseconds = ms;
+     for (SNTStoredEvent *se in events) {
+       se.fileBundleHash = bh;
+       se.fileBundleBinaryCount = @(events.count);
+       se.fileBundleHashMilliseconds = ms;
+     }
+
+     // Send the results to santad. It will decide if they need to be synced.
+     SNTXPCConnection *daemonConn = [SNTXPCControlInterface configuredConnection];
+     [daemonConn resume];
```

Comment:
The connections that AppDelegate local variables so they can be created / released during connection / re-connection attempts. It seems saner to me if we create them as needed. There is a bit more overhead, but seems worth it for cleaner code.